### ISSUE TITLE
Honor gunicorn access log format

### ISSUE THREAD

**Issue by ipmb** [Kludex/uvicorn#527]
When running under gunicorn, [`--access-logformat`](http://docs.gunicorn.org/en/stable/settings.html#access-log-format) isn't honored.

There was some discussion of this in #389, but I didn't see an open issue for it.

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/uvicorn/issues/527">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/uvicorn/issues/527/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/encode/uvicorn/issues/527/pledge.svg">
</picture>
</a>

**Comment by danieltahara** [comment]
This is a pretty big loss for observability for us, and means we have to implement something out of band (and therefore less efficient). Any word on if it will be worked on?

**Comment by loudsquelch** [comment]
Any update on this?

**Comment by asgoel** [comment]
Any update on this one?

**Comment by lovelydinosaur** [comment]
Hey folks - "Any update on this" is almost always not helpful. If there was an update on it, it'd be here on the ticket.
If anyone wants to put some work into progressing it that'd be welcome.

**Comment by asgoel** [comment]
@tomchristie I would be willing to take a stab. i've taken a look at the code and it seems like the `UvicornWorker` is fundamentally written in a different way than the rest of the gunicorn workers that come bundled by default. 

We would basically need to thread through the `Logger` instance from gunicorn and call the `access` method when the request is complete in `httptools_impl`/`h11_impl`. We'd also need to record request time which the app currently does not do. 

Does this approach sound somewhat correct to you?

**Comment by asgoel** [comment]
@tomchristie just following up to see if my above comment made sense.

**Comment by insurandoag** [comment]
We have the same problem at the moment and are about to write a custom logging engine for this, because we must be able to alter the uvicorn log format when running it from gunicorn

**Comment by KlausGlueckert** [comment]
We have the same problem at the moment and are about to write a custom logging engine for this, because we must be able to alter the uvicorn log format when running it from gunicorn. Any update on this if there will be a fix? In addition, ideally it should also be configurable in a logging.conf:

```
[formatter_AccessFormatter]
format=|| %(levelname)s || %(message)s
datefmt=%Y-%m-%dT%H:%M:%S
class=logging.Formatter
access_log_format = %(h)s %(l)s %(u)s %(t)s %(r)s %(s)s %(b)s %(f)s %(a)s
```

**Comment by immerrr** [comment]
I have implemented the gunicorn logging in a very naive and straightforward way in [my branch](https://github.com/encode/uvicorn/compare/master...immerrr:implement-gunicorn-access-log-format).

Would you be interested in a PR?

**Comment by euri10** [comment]
yes @immerrr !
can't promise a speedy speedy review though, pretty bysu

**Comment by immerrr** [comment]
@euri10 sure, np, here you go: https://github.com/encode/uvicorn/pull/947

**Comment by chbndrhnns** [comment]
Now that there is a PR, what would be required to add this functionality? We'd like to match our existing aiohttp-created logs with some new logs created from a FastAPI -> uvicorn -> gunicorn stack.

**Comment by euri10** [comment]
> Now that there is a PR, what would be required to add this functionality? We'd like to match our existing aiohttp-created logs with some new logs created from a FastAPI -> uvicorn -> gunicorn stack.

tl;dr :  not a priority on my side at all

The longer version is I'm lacking time mostly, plus I dont use gunicorn nor am familiar enough with its internals to have a good idea if the implementation is correct or sensible. On top of that there are 0 tests for the feature which makes me very hesitant since this part of the code (UvicornWorker) is already weakly tested in our codebase.

**Comment by immerrr** [comment]
@euri10 the low priority thing is understandable. i asked it in the PR, but let me reiterate, is there anything that would help to get the ball rolling here? 

i mean, if it's the gunicorn part, maybe we can ask nicely in the gunicorn community to have someone come over and have a look. if it's the tests, then i can write them risking that the test code will be thrown away if the approach is not solid enough, but it's fine, i guess. WDYT?

**Comment by euri10** [comment]
I answered in the PR @immerrr

**Comment by lambdaq** [comment]
Can we merge this PR? This is extremely useful for me.

**Comment by KaranTrivedi** [comment]
I just had my issue merged into this thread. Any updates on this?

**Comment by anton-m-py** [comment]
While this is not implemented, [asgi-logger](https://github.com/Kludex/asgi-logger) might be useful to customize access log format.
It supports showing response time.

**Comment by suhcrates-web** [comment]
I got the answer.
see that gunicorn anyway writes log records, and it follows certain default format. And such default format is configured in somwhere. Then, you can modify that config.
 then, where is the default config? it is in the /sitepackages/gunicorn/glogging.py   where the gunicorn is installed. it is the 'access_fmt' variable in the 'Logger' class.

**Comment by pythonwood** [comment]
gunicorn support async base worker offically now. may be is time to rewrite UvicornWorker([AsyncWorker](https://github.com/benoitc/gunicorn/blob/master/gunicorn/workers/base_async.py)).


AsyncWorker source: 

```
class AsyncWorker(base.Worker):
     ...
    def handle_request(self, listener_name, req, sock, addr):
                ...
                request_time = datetime.now() - request_start
                self.log.access(resp, req, environ, request_time)
     
```

it is so long to wait. #606 
may be i will try it

**Comment by waketzheng** [comment]
I am using FastAPI, and this worked for me:
https://github.com/tiangolo/fastapi/issues/1508#issuecomment-638365277

**Comment by ZaneL1u** [comment]
Any update on this?
